WMF Framework/WMFArticle+Migration.swift

```diff
@@ -0,0 +1,392 @@
+
+        if let existing = page.savedInfo {
+            existing.savedDate = snapshot.savedDate
+        } else {
+            let saved = CDPageSavedInfo(context: wmfContext)
+            saved.savedDate = snapshot.savedDate
+            page.savedInfo = saved
+        }
+    }
+
+    private static func makePredicate(for ids: PageIDs) -> NSPredicate {
+        NSPredicate(format: "projectID == %@ AND namespaceID == %d AND title == %@",
+                    ids.projectID, ids.namespaceID, ids.title)
+    }
+
+    // MARK: - Delete all
+
+    private func clearAllSavedData() {
+        guard let wmfDataStore = WMFDataEnvironment.current.coreDataStore else {
+            DDLogError("[SavedPagesMigration] Missing WMFData store")
+            return
+        }
+
+        guard let wmfContext = try? wmfDataStore.newBackgroundContext else {
+            DDLogError("[SavedPagesMigration] Could not create WMFData background context")
+            return
+        }
+        wmfContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+
+        wmfContext.perform {
+            do {
+                let savedInfoFR = NSFetchRequest<NSFetchRequestResult>(entityName: "CDPageSavedInfo")
+                let deleteSavedInfo = NSBatchDeleteRequest(fetchRequest: savedInfoFR)
+                deleteSavedInfo.resultType = .resultTypeObjectIDs
+
+                if let result = try wmfContext.execute(deleteSavedInfo) as? NSBatchDeleteResult,
+                   let deletedIDs = result.result as? [NSManagedObjectID],
+                   !deletedIDs.isEmpty {
+                    NSManagedObjectContext.mergeChanges(
+                        fromRemoteContextSave: [NSDeletedObjectsKey: deletedIDs],
+                        into: [wmfContext, (try? wmfDataStore.viewContext)].compactMap { $0 }
+                    )
+                }
+            } catch {
+                DDLogError("[SavedPagesMigration] Batch clear in WMFData failed: \(error)")
+            }
+        }
+    }
+
+    // MARK: - Helpers / Models
+
+    private struct PageIDs: Sendable, Equatable {
+        let projectID: String
+        let namespaceID: Int16
+        let title: String
+    }
+
+    private struct SavedArticleSnapshot: Sendable, Equatable {
+        let ids: PageIDs
+        let savedDate: Date
+        let viewedDate: Date?
+    }
+
+    private static func pageIDs(from articleURL: URL) -> PageIDs? {
+        guard
+            let siteURL = articleURL.wmf_site,
+            let project = WikimediaProject(siteURL: siteURL),
+            let wmfProject = project.wmfProject
+        else { return nil }
+
+        let title = (articleURL.wmf_title ?? "").normalizedForCoreData
+        let namespaceID = Int16(articleURL.namespace?.rawValue ?? 0)
+        return PageIDs(projectID: wmfProject.id, namespaceID: namespaceID, title: title)
+    }
+
+    private static func getIdFromLegacyArticleURL(_ article: WMFArticle) -> PageIDs? {
```

Comment:
(Optional) Nearly duped from the previous method, might help to use use `pageIDs(from` wherever this is called instead.